clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2015,20 +1956,43 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
   // All special SYCL objects must have __init method. We extract types for
   // kernel parameters from __init method parameters. We will use __init method
   // and kernel parameters which we build here to initialize special objects in
-  // the kernel body. Accessors require additional processing and are handled in
-  // handleSyclAccessorType.
+  // the kernel body.
   bool handleSpecialType(FieldDecl *FD, QualType FieldTy) {
     const auto *RecordDecl = FieldTy->getAsCXXRecordDecl();
-    assert(RecordDecl && "The stream/sampler must be a RecordDecl");
-    CXXMethodDecl *InitMethod = getMethodByName(RecordDecl, InitMethodName);
-    assert(InitMethod && "The stream/sampler must have the __init method");
+    assert(RecordDecl && "The accessor/sampler must be a RecordDecl");
+    llvm::StringLiteral MethodName = KernelDecl->hasAttr<SYCLSimdAttr>()
```

Comment:
Will streams and samplers have `InitESIMDMethodName` method?

Reply:
By my reading of the code, they'll have to, but it'd be good to double-check that's the expectation in reality.  I suspect that's the case as this code looks to have been moved from elsewhere.

Reply:
I need to verify that with Denis.

Reply:
So I got feedback from Denis. ESIMD doesn't support streams and samplers.

Reply:
We have a way to differentiate between stream and accessor/sampler, via the existence of the __finalize method. But there is no way to differentiate between an accessor and a sampler. They both have __init member function. Unless the SYCLSimd attribute is applied only to functions of an accessor?
Can I write something like this:
void test(int val) {
  cl::sycl::sampler smplr;
  kernel_single_task<class first_kernel>([=]() __attribute__((sycl_explicit_simd)) {
    smplr.use();
  });
}
In which case we are in trouble :-) 
@bader, @elizabethandrews, @AaronBallman Any thoughts on this?

Reply:
Talked to Denis about this test case. Although it is legal, the test case will not work because ESIMD functions are not supported by sampler. It is however in the plan for stream and sampler to support ESIMD function. But for now we need to find a way to fix the issue.

Reply:
I do not know if this is valid code. I am not familiar with ESIMD. From a FE perspective, prior to your change this code would work. We handled samplers/streams in ESIMD kernels the same way as non-ESIMD kernels - i.e. looking for `__init` function. With your change, this code will now hit the assert at L1989, since it will look for `__init_esimd` instead now.  

If above code is valid, we need `__init_esimd` for samplers/streams too I think. If above code is invalid, and we want to diagnose it, we need to differentiate accessors from streams and samplers. I do not know required behavior here. @bader

Reply:
According to the proposal we discussed, FE should not diagnose specific SYCL types like accessor, sampler, etc. The FE should validate the contract for new attribute, which is:

1. the attribute is applied to struct/class
2. struct/class must have `__init` member function
3. struct/class may have `__finalize` member function

@againull, BTW, I looked that the `stream` implementation and I'm not sure why do we need `__finalize` member function at all. Can't we do "finalization" in the destructor?

DPC++ library developers are expected to apply this attribute to `accessor`, `sampler`, `stream` and other classes which can't be used as SPIR kernel parameter.

@kbobrovs, @DenisBakhvalov, I guess this should work for ESIMD kernels as well. Am I right?

While I was thinking about additional limitations: we could use "function" attribute instead of class/struct attribute. This should give library developers freedom to use arbitrary name for `__init` method. Sketch of this idea:

```c++
template <...>
class stream {
private:
void stream_init_method(/* parameters to use for SPIR kernel */) __attribute__((spir_kernel_arg_init));
void stream_final_method( ); // if we really need an explicit call of finalization method at the end of kernel execution, there should be one more attribute.
```

OTOH, I'm not sure if using names other than `__init` and `__finalize` is required.

Reply:
> I do not know if this is valid code. I am not familiar with ESIMD. From a FE perspective, prior to your change this code would work. We handled samplers/streams in ESIMD kernels the same way as non-ESIMD kernels - i.e. looking for `__init` function. With your change, this code will now hit the assert at L1989, since it will look for `__init_esimd` instead now.
> 
> If above code is valid, we need `__init_esimd` for samplers/streams too I think. If above code is invalid, and we want to diagnose it, we need to differentiate accessors from streams and samplers. I do not know required behavior here. @bader

@elizabethandrews  without this patch the test case above is crashing here https://github.com/intel/llvm/blob/sycl/clang/lib/Sema/SemaSYCL.cpp#L2338 
So the assertion is going off with and without this patch. So the behavior is essentially the same.

Reply:
> @kbobrovs, @DenisBakhvalov, I guess this should work for ESIMD kernels as well. Am I right?

I don't think ESIMD currently supports using streams. The last time I checked which was ~6 months ago, using streams in an ESIMD kernel resulted in a crash. 
Also, I haven't seen that someone used samplers in ESIMD kernels, so can't comment here.
But accessors should work, and yes, they are distinguished based on `__init_esimd` method (implemented by @kbobrovs).
So, as far as there are no ESIMD regressions, I think it's OK from my side.

Reply:
@kbobrovs, what is the motivation to have `__init_esimd`? It seems to match `__init` for all classes except accessor, where `__init_esimd` initializes only accessor pointer, but leave other accessor fields uninitialized, which seems like great source of hard to debug issues. If other fields are not initialized, I expect kernel argument elimination pass to remove unused accessor fields only they are not used (i.e. does the same as `__init_esimd`, but in a safe way).
I suggest drop `__init_esimd` and use `__init`. What do you think?

Reply:
> @elizabethandrews without this patch the test case above is crashing here https://github.com/intel/llvm/blob/sycl/clang/lib/Sema/SemaSYCL.cpp#L2338
> So the assertion is going off with and without this patch. So the behavior is essentially the same.

Thanks for clarifying. I guess there is an existing bug in code then i.e. ArgsSizeChecker crashes because of the same concern I mention here (__init/__init_esimd not handled correctly for special classes). Ideally the compiler should not be crashing in this case either but I guess that can be done as follow-up if other reviewers are ok with it. Please file an issue to keep track of it, if you do it as follow-up

Reply:
Accessors used in ESIMD were restricted compared to normal SYCL accessors to match the "surfaces" used in CM. For example, they did not support the access range and access offset, and we could save on the number of arguments passed to ESIMD kernels. Those restrictions were supposed to be diagnosed, but this hasn't been implemented.

Overall, I agree with the idea of dropping __init_simd and going the generic way, as we have the unused argument optimization now.

Reply:
I'd expect parameter creation for streams and samplers to work the same way as for vanilla SYCL. The crash can be caused by BE issues, as these have never been tested and officially supported by ESIMD.

Reply:
> Overall, I agree with the idea of dropping __init_simd and going the generic way, as we have the unused argument optimization now.

One note: the optimization is off default in `sycl` branch and enabled with `-fsycl-dead-args-optimization`.
@mdtoguchi has a PR enabling it by default here: https://github.com/intel/llvm/pull/3004.

> I'd expect parameter creation for streams and samplers to work the same way as for vanilla SYCL. The crash can be caused by BE issues, as these have never been tested and officially supported by ESIMD.

If these features are used, but not supported in ESIMD mode, the compiler should give an error instead of crashing. Right?

Reply:
Streams should work with ESIMD too. This needs more investigation.